src/mpi/init/init_async.c

```diff
@@ -59,22 +74,163 @@ static void progress_fn(void *data)
+{
+    int th_idx, read_count = 0, mpi_errno = MPI_SUCCESS;
+    char *affinity_copy = NULL;
+    const char *affinity_to_parse = MPIR_CVAR_CH4_PROGRESS_THREAD_AFFINITY;
+    char *proc_id_str, *tmp;
+    size_t proc_count;
+
+    if (!affinity_to_parse || strlen(affinity_to_parse) == 0) {
+        printf("ERROR: unexpected affinity string\n");
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+    /* create copy of original buffer because it will be modified in strsep */
+    affinity_copy = MPL_strdup(affinity_to_parse);
+    MPIR_Assert(affinity_copy);
+    tmp = affinity_copy;
+
+    /* apply auto affinity */
+    if (!strcmp(affinity_copy, "auto")) {
+        /* generate default affinity */
+        proc_count = MPL_get_nprocs();
+        for (th_idx = 0; th_idx < async_threads_per_node; th_idx++) {
+            if (th_idx < proc_count)
+                thread_affinity[th_idx] = proc_count - th_idx - 1;
+            else
+                thread_affinity[th_idx] = thread_affinity[th_idx % proc_count];
+        }
+        goto fn_exit;
+    }
+
+    /* apply explicit affinity to the provided logical cpu cores */
+    for (th_idx = 0; th_idx < async_threads_per_node; th_idx++) {
+        proc_id_str = MPL_strsep(&tmp, ",");
+        if (proc_id_str != NULL) {
+            if ((strlen(proc_id_str) > 0 && !isdigit(proc_id_str[0])) || atoi(proc_id_str) < 0) {
+                printf
+                    ("ERROR: unexpected proc_id %s, affinity string %s\n",
+                     proc_id_str, affinity_to_parse);
+                mpi_errno = MPI_ERR_OTHER;
+                goto fn_fail;
+            }
+            thread_affinity[th_idx] = atoi(proc_id_str);
+            read_count++;
+        } else {
+            printf
+                ("ERROR: unexpected end of affinity string, expected %d numbers, read %d, affinity string %s\n",
+                 async_threads_per_node, read_count, affinity_to_parse);
+            mpi_errno = MPI_ERR_OTHER;
+            goto fn_fail;
+        }
+    }
+    if (read_count < async_threads_per_node) {
+        printf
+            ("ERROR: unexpected number of processors (specify 1 logical processor per 1 progress thread), affinity string %s\n",
+             affinity_to_parse);
+        mpi_errno = MPI_ERR_OTHER;
+        goto fn_fail;
+    }
+
+  fn_exit:
+    MPL_free(affinity_copy);
+    return mpi_errno;
+
+  fn_fail:
+    goto fn_exit;
+}
+#endif
+
 /* called inside MPID_Init_async_thread to provide device override */
 int MPIR_Init_async_thread(void)
 {
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno = MPI_SUCCESS, thr_err;
+    int global_rank, local_rank, local_size, async_threads_per_node;
+    int *thread_affinity = NULL;
+    bool apply_affinity;
+    int have_cliques, affinity_idx;
     MPIR_FUNC_TERSE_STATE_DECL(MPID_STATE_MPIR_INIT_ASYNC_THREAD);
 
     MPIR_FUNC_TERSE_ENTER(MPID_STATE_MPIR_INIT_ASYNC_THREAD);
 
+#if defined(MPL_HAVE_MISC_GETNPROCS)
+
```

Comment:
Can we wrap the following into a static function?

Reply:
Sure. Do you think it's okay to add an extra refactor patch to add such wrapper?

Reply:
You can do that. But we need to cleanup (squash) the commits for final review.

Reply:
Added a wrapper.